game/src/main/java/org/apollo/game/io/player/JdbcPlayerSerializer.java

```diff
@@ -1,33 +1,464 @@
+import org.apollo.game.model.Position;
 import org.apollo.game.model.World;
 import org.apollo.game.model.entity.Player;
+import org.apollo.game.model.entity.Skill;
+import org.apollo.game.model.entity.SkillSet;
+import org.apollo.game.model.entity.attr.*;
+import org.apollo.game.model.entity.setting.Gender;
+import org.apollo.game.model.entity.setting.PrivilegeLevel;
+import org.apollo.game.model.inv.Inventory;
+import org.apollo.game.model.inv.SlottedItem;
+import org.apollo.net.codec.login.LoginConstants;
 import org.apollo.util.security.PlayerCredentials;
 
+import java.sql.*;
+import java.time.LocalDateTime;
+import java.util.Map;
+
+import static java.util.Objects.requireNonNull;
+
 /**
  * A {@link PlayerSerializer} that utilises {@code JDBC} to communicate with an SQL database containing player data.
  *
  * @author Major
+ * @author Sino
  */
+// TODO optimize use of collections to avoid boxing by using fastutil
 public final class JdbcPlayerSerializer extends PlayerSerializer {
+	/**
+	 * Maps skill id to the enum value that is known by the database
+	 * and back.
+	 */
+	private static final ImmutableBiMap<Integer, String> SKILL_ENUMS = ImmutableBiMap.<Integer, String>builder()
+			.put(Skill.ATTACK, "attack")
+			.put(Skill.STRENGTH, "strength")
+			.put(Skill.DEFENCE, "defence")
+			.put(Skill.HITPOINTS, "hitpoints")
+			.put(Skill.RANGED, "ranged")
+			.put(Skill.PRAYER, "prayer")
+			.put(Skill.MAGIC, "magic")
+			.put(Skill.COOKING, "cooking")
+			.put(Skill.FISHING, "fishing")
+			.put(Skill.WOODCUTTING, "woodcutting")
+			.put(Skill.FIREMAKING, "firemaking")
+			.put(Skill.FLETCHING, "fletching")
+			.put(Skill.FARMING, "farming")
+			.put(Skill.SLAYER, "slayer")
+			.put(Skill.AGILITY, "agility")
+			.put(Skill.CRAFTING, "crafting")
+			.put(Skill.HERBLORE, "herblore")
+			.put(Skill.THIEVING, "thieving")
+			.put(Skill.MINING, "mining")
+			.put(Skill.SMITHING, "smithing")
+			.put(Skill.RUNECRAFT, "runecraft")
+			.build();
+
+	/**
+	 * Maps {@link PrivilegeLevel} to the enum value that is known by the
+	 * database and back.
+	 */
+	private static final ImmutableBiMap<PrivilegeLevel, String> RANK_ENUMS = ImmutableBiMap.<PrivilegeLevel, String>builder()
+			.put(PrivilegeLevel.STANDARD, "player")
+			.put(PrivilegeLevel.MODERATOR, "moderator")
+			.put(PrivilegeLevel.ADMINISTRATOR, "administrator")
+			.build();
+
+	/**
+	 * Maps {@link AttributeType}s to the enum value that is known by the
+	 * database and back.
+	 */
+	private static final ImmutableBiMap<AttributeType, String> ATTRIBUTE_TYPE_ENUMS = ImmutableBiMap.<AttributeType, String>builder()
+			.put(AttributeType.STRING, "string")
+			.put(AttributeType.BOOLEAN, "boolean")
+			.put(AttributeType.DOUBLE, "double")
+			.put(AttributeType.LONG, "long")
+			.build();
+
+	/**
+	 * The associated queries that are ran by this serializer.
+	 */
+	private static final String
+			GET_ACCOUNT_QUERY = "SELECT password_hash, rank FROM get_account(?::text)",
```

Comment:
Why are these queries implemented as functions?

Reply:
Because majority of them are slightly more complex than the usual one liners involving joins and such. Java doesn't support text blocks until 13 and we're on 11 now so that'd get a bit messy. Do you have a better idea?

Reply:
If the only reason is to cut down on the size of string literals than we can store the queries in a kotlin file (which does support multi-line raw literals). The game module already has the Kotlin compiler enabled so this won't be an issue.